File: backend/controllers/web_auth_controllers.go
package controllers

import (
	//"fmt" // for debugging

	"context"
	"fmt"
	"net/http"
	"os" // to access .env file

	"eunity.com/backend-main/helpers/DBManager"
	"eunity.com/backend-main/helpers/SessionManager"
	"eunity.com/backend-main/models"
	"github.com/gin-gonic/gin"
	"github.com/joho/godotenv" // to load .env file
	"github.com/markbates/goth"
	"github.com/markbates/goth/gothic"
	"github.com/markbates/goth/providers/google"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	//"go.mongodb.org/mongo-driver/bson"
	//"go.mongodb.org/mongo-driver/bson/primitive"
)

//we will be using this endpoints for login into the dashboard

type Web_Auth_controllers struct{}

// required google env variables
var GoogleClientID string
var GoogleClientSecret string

// var GoogleRedirectURI string = "http://localhost:3200/api/v1/webAuth/google/callback"
var GoogleRedirectURI string = "https://eunityusa.com/api/v1/webAuth/google/callback"

func init() {

	godotenv.Load() //loads the .env file
	GoogleClientID = os.Getenv("GOOGLE_CLIENT_ID")
	GoogleClientSecret = os.Getenv("GOOGLE_SECRET")

	fmt.Println(GoogleClientID)
	fmt.Println(GoogleClientSecret)

	scopes := []string{
		"email",
		"profile",
	}

	goth.UseProviders(
		google.New(GoogleClientID, GoogleClientSecret, GoogleRedirectURI, scopes...),
	)
}

// @Summary Begin Google Auth
// @Description Begins the google auth process
// @Tags Web Auth
// @Accept json
// @Produce json
// @Success 200 {string} string "Google Auth Started"
// @Router /webAuth/google [get]
func (ac *Web_Auth_controllers) GET_BeginGoogleAuth(c *gin.Context) {

	//request quert handler object
	q := c.Request.URL.Query()

	//add this to the query
	q.Add("provider", "google")

	//encode the query
	c.Request.URL.RawQuery = q.Encode()

	//begin the auth handler with gothic
	gothic.BeginAuthHandler(c.Writer, c.Request)
}

// @Summary Google OAuth Callback
// @Description Callback for google auth
// @Tags Web Auth
// @Accept json
// @Produce json
// @Success 200 {string} string "Google Auth Callback"
// @Router /webAuth/google/callback [get]
func (ac *Web_Auth_controllers) GET_GoogleOAuthCallback(c *gin.Context) { // #TODO clean this up - @AggressiveGas
	q := c.Request.URL.Query()
	q.Add("provider", "google")
	c.Request.URL.RawQuery = q.Encode()
	//view entire query data

	user, err := gothic.CompleteUserAuth(c.Writer, c.Request)
	if err != nil {
		c.AbortWithError(http.StatusInternalServerError, err)
		return
	}

	//check if the developer is already in the database
	email := user.Email

	dev := DBManager.DB.Collection("whitelisted-devs").FindOne(context.Background(), bson.M{"email": email})
	if dev.Err() != nil {
		c.JSON(400, gin.H{
			"response": "Account not authorized",
		})
	}

	//check if they exists in the developer collection if net we add all their info
	//to the developer collection
	developer := DBManager.DB.Collection("developers").FindOne(context.Background(), bson.M{"email": email})
	fmt.Println(email)
	fmt.Println(user.FirstName)
	fmt.Println(user.LastName)
	fmt.Println(user.Name)
	fmt.Println(user.RawData["verified_email"])
	fmt.Println(user.UserID)

	objectID := primitive.NewObjectID()

	if developer.Err() != nil {
		newDev := models.Developer{
			ID:        &objectID,
			Email:     email,
			FirstName: user.FirstName,
			LastName:  user.LastName,
			AvatarURL: user.AvatarURL,
			Providers: map[string]models.Provider{
				"google": {
					Name:           user.Name,
					Email:          user.Email,
					Email_verified: user.RawData["verified_email"].(bool),
					Sub:            user.UserID,
				},
			},
		}

		fmt.Println("made it here bruh")

		//create a new developer
		_, err := DBManager.DB.Collection("developers").InsertOne(context.Background(), newDev)
		if err != nil {
			c.JSON(400, gin.H{
				"response": "Unable to create developer",
			})
			return
		}
	}

	fmt.Println("made it here bruh 2")

	developer = DBManager.DB.Collection("developers").FindOne(context.Background(), bson.M{"email": email})
	if developer.Err() != nil {
		c.JSON(400, gin.H{
			"response": "Unable to login",
		})
		return
	}

	var result models.Developer
	err = developer.Decode(&result)
	if err != nil {
		c.JSON(400, gin.H{
			"response": "Unable to login",
		})
		return

	}

	_, err = SessionManager.Create_Developer_Session(result.ID.Hex(), c)
	if err != nil {
		c.JSON(400, gin.H{
			"response": "Unable to login",
		})
		return

	}

	//after this we want to create a session

	//we need to implement the rest for web auth here we would create the user etc, etc
	//redirect to /dashboard should have access to all the developer endpoints
	c.Redirect(http.StatusMovedPermanently, "/dashboard")

}

File: python/controllers/web_auth_controllers.py
from fastapi import FastAPI, HTTPException, Depends
from fastapi.responses import RedirectResponse
from fastapi.security import OAuth2PasswordBearer
from pydantic import BaseModel
from typing import Optional

import os
import asyncio
from bson.objectid import ObjectId
from pymongo import MongoClient

app = FastAPI()

# Load .env file
if os.path.exists(".env"):
    with open(".env", "r") as f:
        for line in f.readlines():
            key, value = line.strip().split("=")
            os.environ[key] = value

# MongoDB connection
client = MongoClient(os.getenv("MONGO_DB_URI"))
db = client["eunity-main"]

# Google OAuth settings
GOOGLE_CLIENT_ID = os.getenv("GOOGLE_CLIENT_ID")
GOOGLE_CLIENT_SECRET = os.getenv("GOOGLE_CLIENT_SECRET")
GOOGLE_REDIRECT_URI = os.getenv("GOOGLE_REDIRECT_URI")

# OAuth2 scheme
oauth2_scheme = OAuth2PasswordBearer(tokenUrl="/token")


class User(BaseModel):
    email: str
    first_name: str
    last_name: str
    avatar_url: str
    user_id: str


class Developer(BaseModel):
    id: ObjectId
    email: str
    first_name: str
    last_name: str
    avatar_url: str
    providers: dict


async def get_current_user(token: str = Depends(oauth2_scheme)):
    # Implement token verification and user lookup logic here
    pass


@app.get("/webAuth/google")
async def begin_google_auth():
    # Begin Google auth process
    scopes = ["email", "profile"]
    auth_url = f"https://accounts.google.com/o/oauth2/auth?client_id={GOOGLE_CLIENT_ID}&redirect_uri={GOOGLE_REDIRECT_URI}&scope={'+'.join(scopes)}&response_type=code"
    return RedirectResponse(url=auth_url, status_code=302)


@app.get("/webAuth/google/callback")
async def google_oauth_callback(code: str):
    # Handle Google OAuth callback
    user, err = await complete_user_auth(code)
    if err:
        raise HTTPException(status_code=400, detail="Unable to authenticate")

    email = user.email
    dev = db["whitelisted-devs"].find_one({"email": email})
    if dev is None:
        return {"response": "Account not authorized"}

    developer = db["developers"].find_one({"email": email})
    if developer is None:
        new_dev = Developer(
            email=user.email,
            first_name=user.first_name,
            last_name=user.last_name,
            avatar_url=user.avatar_url,
            providers={
                "google": {
                    "name": user.name,
                    "email": user.email,
                    "email_verified": user.raw_data["verified_email"],
                    "sub": user.user_id,
                }
            },
        )
        db["developers"].insert_one(new_dev)

    developer = db["developers"].find_one({"email": email})
    session_id = await create_developer_session(developer["_id"], " dashboard")
    return RedirectResponse(url="/dashboard", status_code=302)


async def complete_user_auth(code: str):
    # Implement Google OAuth authentication logic here
    pass


async def create_developer_session(dev_id: str, redirect_url: str):
    # Implement session creation logic here
    pass
